src/mem/arena.rs

```diff
@@ -204,7 +207,7 @@ impl<'arena, A: Arena + ?Sized> ArenaExt<'arena> for &'arena A {
 /// arena.reset();
 /// let buf3 = arena.alloc_slice::<u8>(64)?;
 /// assert_eq!(buf3.len(), 64);
-/// # Ok::<(), OutOfMemory>(())
+/// # Ok::<(), manticore::Error<OutOfMemory>>(())
```

Comment:
Personally I actually don't think this is a very useful error: getting an error location from a failed allocation will point inside of arena code, when it would be much more useful for the caller to see the arena call that emitted the error. WDYT?

Reply:
A follow-on task I'm envisioning (and that you incepted me with a few months ago 😄) is to steal ideas from https://github.com/rust-lang/rust/issues/87401: investigate whether we can use `#[track_caller]` to automatically track the locations of error conversions performed via `?`. That'd give extra, more useful debug information and would save the need to add `fail!` calls in many places.

So my thinking is:
1. Use the `Error` type with the `Arena` trait i.e. no change to current PR state.
2. If the debug information printed is not useful, callers can still manually add debug info. This is actually already done in some places in src/io/read.rs:
    ```rust
            let out = arena
                .alloc_raw(layout)
                .map_err(|_| fail!(io::Error::BufferExhausted))?;
    ```
    Since adding the fallible `Arena` trait calls are only made Since the `Arena` trait
3. Add error conversion tracking using `#[track_caller]` if possible. If it isn't, rely on step 2.

FWIW switching the `Arena` trait to using `core::result::Result` would be pretty easy and I can do that if you do feel strongly that seeing where in arena.rs the error originated would never be useful.

Reply:
I don't feel strongly, if you plan on doing stuff with track_caller.